ENH: Vectorize quicksort for 32-bit dtype using AVX-512

r-devulap opened the pull request:
AVX-512 speeds up quicksort by nearly **14x** for uniformly distributed random numbers. The ideas and code for this patch are based on these two research papers:

1. Fast and Robust Vectorized In-Place Sorting of Primitive Types https://drops.dagstuhl.de/opus/volltexte/2021/13775/

2. A Novel Hybrid Quicksort Algorithm Vectorized using AVX-512 on Intel Skylake https://arxiv.org/pdf/1704.08579.pdf

High level idea: Vectorize the quicksort inline partitioning using AVX-512 _compressstore_ instructions. The pivot value is the median of 72 array elements chosen at random. If the array size is < 128, then use Bitonic sorting networks on ZMM registers which can parallelize a lot of the compare and swap required in a sorting algorithm. Good resource for bitonic sorting network: http://mitp-content-server.mit.edu:18180/books/content/sectbyfn?collid=books_pres_0&fn=Chapter%2027.pdf&id=8030

Benchmarking details: 

```
-      52.6±0.1μs       36.2±0.4μs     0.69  bench_function_base.Sort.time_sort('quick', 'uint32', ('ordered',))
-     67.3±0.03μs       35.4±0.4μs     0.53  bench_function_base.Sort.time_sort('quick', 'int32', ('ordered',))
-     85.8±0.03μs       39.9±0.3μs     0.46  bench_function_base.Sort.time_sort('quick', 'float32', ('ordered',))
-     84.0±0.03μs       36.5±0.4μs     0.43  bench_function_base.Sort.time_sort('quick', 'uint32', ('reversed',))
-      105±0.06μs       35.7±0.3μs     0.34  bench_function_base.Sort.time_sort('quick', 'int32', ('reversed',))
-       141±0.1μs       40.4±0.4μs     0.29  bench_function_base.Sort.time_sort('quick', 'float32', ('reversed',))
-       285±0.3μs       38.3±0.2μs     0.13  bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 1000))
-       377±0.3μs       42.7±0.1μs     0.11  bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 100))
-      325±0.07μs       36.6±0.2μs     0.11  bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 1000))
-       370±0.1μs       40.4±0.2μs     0.11  bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 1000))
-       384±0.1μs       38.0±0.2μs     0.10  bench_function_base.Sort.time_sort('quick', 'uint32', ('sorted_block', 10))
-       416±0.3μs       38.3±0.1μs     0.09  bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 100))
-       458±0.2μs       41.0±0.6μs     0.09  bench_function_base.Sort.time_sort('quick', 'uint32', ('random',))
-      476±0.09μs       42.1±0.1μs     0.09  bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 100))
-       424±0.2μs      36.1±0.05μs     0.09  bench_function_base.Sort.time_sort('quick', 'int32', ('sorted_block', 10))
-       487±0.2μs       40.7±0.2μs     0.08  bench_function_base.Sort.time_sort('quick', 'float32', ('sorted_block', 10))
-       495±0.2μs       37.9±0.1μs     0.08  bench_function_base.Sort.time_sort('quick', 'int32', ('random',))
-     61.5±0.02μs       4.69±0.2μs     0.08  bench_function_base.Sort.time_sort('quick', 'uint32', ('uniform',))
-       574±0.2μs       41.7±0.6μs     0.07  bench_function_base.Sort.time_sort('quick', 'float32', ('random',))
-      95.0±0.3μs       4.65±0.1μs     0.05  bench_function_base.Sort.time_sort('quick', 'int32', ('uniform',))
-       139±0.1μs       5.27±0.2μs     0.04  bench_function_base.Sort.time_sort('quick', 'float32', ('uniform',))
```

Commit:
ENH: Use AVX-512 for float quicksort

Commit:
BENCH: Add benchmarks for 32-bit data type sorting

Commit:
TST: Add unit tests for vectorized quicksort

Commit:
ENH: Vectorize sort for int32 and uint32 using AVX-512

Commit:
MAINT: Adding comments

Commit:
MAINT: Fix linter errors

Commit:
MAINT: Workaround for gcc missing intrinsic: use loadu_si512 instead of loadu_epi32

Commit:
TST: Use default_rng to generate random numbers for the test

Commit:
MAINT: Use _mm512_cmp_ps_mask instead of _mm512_cmpunord_ps_mask

Workaround for missing intrinsic in older gcc

seiko2plus commented:
@r-devulap, we should no longer accept any raw SIMD code, only universal intrinsics along with the new dispatch mechanism any missing intrinsics can be emulated by universal intrinsics itself so we can tweak them later and move them to main interface headers. I can offer suggestions for a design/code that can fit universal intrinsics, or just tell me your reasons/concerns that made you change course so we can find a solution together.

r-devulap commented:
>  I can offer suggestions for a design/code that can fit universal intrinsics, or just tell me your reasons/concerns that made you change course so we can find a solution together.

There are a couple of things we might need to address before we re-write this patch using the universal intrinsics framework: 

1) Performance for this patch really hinges on register vector size and the use `compressstore` instruction. As a demonstration of that, I re-wrote the quick sort partitioning function using SSE while emulating the `compressstore` instruction and continuing to use ZMM registers for bitonic sorting network (see https://github.com/r-devulap/numpy/commit/535a10168f6bd28e6db6199c3f88a12b9d3e5020). It wipes out all the performance gains and barely improves over the scalar quicksort (see numbers below). Of course, we would have to re-write Bitonic sorting and the algorithm to pick the pivot element using XMM registers but that would only slow it down further. I am pretty sure that at least for 16 byte registers (SSE, NEON and PowerPC), vectorized sorting would make performance worse. 

```
       before           after         ratio
     [ba366476]       [6c60365d]
     <main>           <sse-sort>
-       494±0.3μs        460±0.2μs     0.93  bench_function_base.Sort.time_sort('quick', 'int32', ('random',))
-       572±0.2μs        463±0.5μs     0.91  bench_function_base.Sort.time_sort('quick', 'float32', ('random',))
```

2) The implementation of Bitonic sorting network is very much dependent on the register size. For example, the [sorting network to sort 16 elements](https://en.wikipedia.org/wiki/Bitonic_sorter#/media/File:BitonicSort.svg) would need just one ZMM register but will need 4 registers in SSE and hence 4 times more shuffles and compares. The SSE/AVX implementation of most of the code used in this patch would be vastly different and would not conform to Universal Intrinsic framework. While there is no problem with having different functions for each SIMD type, but again, I am not sure they will improve performance and hence not worth having.

r-devulap commented:
ping, any suggestions on converting to universal intrinsics format?

r-devulap commented:
ping ..

r-devulap commented:
> This has a merge conflict. Could you also either link to the SSE comment above in the code or copy the content into the header? The information there is valuable.

Fixed the merge conflict and added a comment as well.

r-devulap commented:
I am guessing the failures have nothing to do with the patch? 

`E               subprocess.CalledProcessError: Command '['nm.exe', '--help']' returned non-zero exit status 4294967295.`

mattip commented:
@seiko2plus any thoughts? I would like to put this in as-is, then we can think about replacing the AVX512-specific code with universal intrinsics

seiko2plus commented:
we can accept it replace it later with universal intrinsics but this code uses compiler attribute target rather than the new dispatcher, we had discussed this before within SVML pr. again old dispatching mechanism should not be used anymore, the new dispatcher is more efficient supports all kinds of compilers, and it can be managed through `--cpu-baseline` and `--cpu-dispatch`.

mattip commented on numpy/core/src/npysort/quicksort.c.src:
```diff
@@ -83,11 +85,21 @@
+    }
+#endif
+#endif
+
```
@seiko2plus you mean this should be removed and the code should use the new dispatcher semantics?

seiko2plus replied:
Yes but first we need a dispatch-able source for the avx512_skx kernels or x86 features in general. in other words, `qsort-32bit-avx512.h.src` should be renamed to `x86-qsort.dispatch.c.src`. also, we need another header or any other existence header to hold the forward declarations of the exported functions. for example:

> 1. `x86-qsort.h` (forward declarations)

```C
#include "npy_cpu_dispatch.h"

#ifndef NPY_DISABLE_OPTIMIZATION
    #include "x86-qsort.dispatch.h"
#endif
NPY_CPU_DISPATCH_DECLARE(NPY_NO_EXPORT void x86_quicksort_int,
     (void *start, npy_intp num))

NPY_CPU_DISPATCH_DECLARE(NPY_NO_EXPORT void x86_quicksort_uint,
     (void *start, npy_intp num))
```

> 2.`x86-qsort.dispatch.c.src` (kernels)

```C
/*@targets
 * $maxopt $keep_baseline avx512_skx
 */
// policy $keep_baseline is used to avoid skip building avx512_skx
// when its part of baseline features (--cpu-baseline), since
// 'baseline' option isn't specified within targets.

#include "x86-qsort.h"

#ifdef NPY_HAVE_AVX512_SKX
NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(x86_quicksort_int)
(void *start, npy_intp num)
{
    // avx512 simd code for unsinged int
}

NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(x86_quicksort_uint)
(void *start, npy_intp num)
{
    // avx512 simd code for singed int
}
#endif
```
> 3. runtime dispatch

```diff
+#include "x86-qsort.h"
-#if defined -HAVE_ATTRIBUTE_TARGET_AVX512_SKX_WITH_INTRINSICS
-    if(NPY_CPU_HAVE(AVX512_SKX)) {
-        avx512_qsort_@suff@(start, num);
-        return 0;
-    }
-#endif
+#ifndef NPY_DISABLE_OPTIMIZATION
+    #include "x86-qsort.dispatch.h"
+#endif
+void (*dispfunc)(void*, npy_intp) = NULL;
+NPY_CPU_DISPATCH_CALL_XB(dispfunc = &x86_quicksort_@suff@);
+if (dispfunc) {
+   (*dispfunc)(start, num);
+   return 0;
+}
```
NOTE: `NPY_CPU_DISPATCH_CALL_XB` is no longer accept goto or return since #17201, while makes it easier for readability but as u can realize I had to add extra unnecessary lines to get the 'return' back (compiler going to optimize it out anyway).
before #17201, the runtime-dispatch code can be simplified as follows:
```C
#ifndef NPY_DISABLE_OPTIMIZATION
    #include "x86-qsort.dispatch.h"
#endif
NPY_CPU_DISPATCH_CALL_XB(x86_quicksort_@suff@, (start, num), return 0);
```

r-devulap replied:
Thanks for spelling it out :) Would have found it hard to figure out on my own.

r-devulap commented:
> we had discussed this before within SVML pr. again old dispatching mechanism should not be used anymore, the new dispatcher is more efficient supports all kinds of compilers, and it can be managed through `--cpu-baseline` and `--cpu-dispatch`.

Yeah sorry about that. Old habits die hard. Will replace it with the new dispatcher.

seiko2plus commented:
@r-devulap, an example here https://github.com/numpy/numpy/pull/20133#discussion_r785475466, since we didn't face this case before.

Commit:
MAINT: Use new CPU dispatch mechanism

r-devulap commented:
This patch now uses the new dispatch mechanism.

r-devulap commented:
@seiko2plus does this looks good now?

mattip commented:
@r-devulap just checking that you reran benchmarks and the dispatching works as expected?

If so I think this is good to go, we can iterate on missing bits going forward.

r-devulap commented:
@mattip I am unable to run any bench-compare and keep getting this error when I run `python runtests.py --bench-compare main`: 

```
   STDOUT -------->

   STDERR -------->
   Running from numpy source directory.
   Traceback (most recent call last):
     File "setup.py", line 89, in <module>
       raise RuntimeError(
   RuntimeError: Setuptools version is '60.2.0', version < '60.0.0' is required. See pyproject.toml
```

`pip list` tells me my `setuptools` is at `59.2.0`. Any idea what could be wrong?

mattip commented:
Does the benchmark runner set up a virtualenv in the .benchmark directory?

r-devulap commented:
I think it tries to do that in `benchmarks/env/`

`numpy/benchmarks/env/5123c33b77ca2d1538faa7bba52fb56a/bin/python setup.py build (exit status 1)`

seiko2plus commented on numpy/core/src/npysort/x86-qsort.dispatch.c.src:
```diff
@@ -0,0 +1,585 @@
+    if(right - left == 16) {
+        @zmm_t@ vec = _mm512_loadu_@vsuf3@(arr + left);
+        npy_int amount_gt_pivot = partition_vec_@vsuf1@(arr, left, left + 16, vec, pivot_vec, &min_vec, &max_vec);
+        *smallest = _mm512_reduce_min_@vsuf1@(min_vec);
+        *biggest = _mm512_reduce_max_@vsuf1@(max_vec);
```
```suggestion
#ifdef NPY_HAVE_AVX512F_REDUCE
        *smallest = _mm512_reduce_min_@vsuf1@(min_vec);
        *biggest = _mm512_reduce_max_@vsuf1@(max_vec);
#else
        // emulate 
#endif
```
reduce intrinsics aren't backed by native support and not emulated by all compilers that support , see

https://github.com/numpy/numpy/blob/c65bc212ec1987caefba0ea7efe6a55803318de9/numpy/distutils/checks/extra_avx512f_reduce.c#L2-L7

r-devulap replied:
Done. Added it to SIMD module.

seiko2plus commented:
> @seiko2plus does this looks good now?

yes, I think so.

seiko2plus commented:
@r-devulap, then upgrade setuptools via pip or manually choose the python version you use through config file:
https://github.com/numpy/numpy/blob/c65bc212ec1987caefba0ea7efe6a55803318de9/benchmarks/asv_compare.conf.json.tpl#L40-L42

you can also provide the path of python exec manually within:
https://github.com/numpy/numpy/blob/c65bc212ec1987caefba0ea7efe6a55803318de9/benchmarks/asv_compare.conf.json.tpl#L79-L84

Commit:
MAINT: Remove unnecessary function optimize attribute

goldsteinn commented:
Hi,

I think there might be an off by 1 bug in the right/left/pivot index computation.

built the avx512 sort functions as follows:

```
$> python3 numpy/distutils/conv_template.py numpy/core/src/npysort/qsort-32bit-avx512.h.src;  gcc -DHAVE_ATTRIBUTE_TARGET_AVX512_SKX_WITH_INTRINSICS -Inumpy/random -Inumpy/random/src -Inumpy/random/src/legacy -Inumpy/core/include -Inumpy/core/include/numpy -Ibuild/src.linux-x86_64-3.8/numpy/distutils/include -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -Inumpy/core/src/_simd -I/usr/include/python3.8 -Inumpy/core/src/common -Inumpy/core/src/npymath -E numpy/core/src/npysort/qsort-32bit-avx512.h > npy-sort-avx512.h
```

I replace the `2*log2(arrsize)` init `max_iters` with `arrsize` and still see `heapsort_uint` being called.

It seems to be looping in a case where `pivot == smallest`. Its not a correctness bug given `max_iters`
but probably a performance issue.

Repo with a reproduction:
https://github.com/goldsteinn/possible-off-by-one

r-devulap commented:
@goldsteinn That isn't a bug and is expected behavior. If you look at the `get_pivot` function, it returns the value of the pivot rather its index (keeping track of index of median of 72 requires additional cpu cycles). It is possible for arrays with skewed distribution of numbers, the pivot value doesn't partition the array and hence qsort won't make progress, which is what is happening with your example. The switch to heapsort exists exactly for this reason. I would think scalar quicksort would have this issue too. The difference with scalar quicksort would be that every step will result in at least one element (the pivot itself) to be partitioned which won't be the case here (because we don't know the index of the pivot).

goldsteinn commented:
@r-devulap 

Hi,

Okay, although "skewed distribution" is slightly misleading. I think limited range is more accurate. With small size ranges
being sorted this case starts to becomes pretty likely.

Out of curiosity would it be possible call `heapsort` in the case where `pivot_index` is equal to `smallest` or `biggest`? Or is there a case where the next recursive call it will actually make progress?

Commit:
SIMD: Add intrinsics emulation for reduce_min and reduce_max instructions

r-devulap commented on numpy/core/src/npysort/x86-qsort.dispatch.c.src:
```diff
@@ -0,0 +1,585 @@
+    a = _mm256_min_@vsuf1@(a, b);                                         \
+    b = _mm256_max_@vsuf1@(temp, b);}                                     \
+
+static NPY_INLINE NPY_GCC_OPT_3
```
Done.

r-devulap commented:
> Out of curiosity would it be possible call `heapsort` in the case where `pivot_index` is equal to `smallest` or `biggest`? Or is there a case where the next recursive call it will actually make progress?

I think this case already handled. if `pivot == smallest`, then we don't even need to sort the left array (because they are all of same value) and similarly if `pivot == largest`, then we don't need to sort the right array. Right?

goldsteinn commented:
> > Out of curiosity would it be possible call `heapsort` in the case where `pivot_index` is equal to `smallest` or `biggest`? Or is there a case where the next recursive call it will actually make progress?
> 
> I think this case already handled. if `pivot == smallest`, then we don't even need to sort the left array (because they are all of same value) and similarly if `pivot == largest`, then we don't need to sort the right array. Right?

Whats happening though is the same pivot is being returned each time when neither `smallest` nor `biggest` are updated. So its just a really expensive path to `heapsort` (and causes bad performance for dense ranges).

Commit:
MAINT: Use cvt instruction instead of extract

r-devulap commented:
> Whats happening though is the same pivot is being returned each time when neither `smallest` nor `biggest` are updated. So its just a really expensive path to `heapsort` (and causes bad performance for dense ranges).

Do you have an example of an array where this worsens performance? At least for the example you provided in https://github.com/goldsteinn/possible-off-by-one/blob/master/test.c, this patch still speeds up quicksort by nearly 5x. See the benchmark: https://github.com/r-devulap/numpy/commit/aaf8213d729798eb9b3356ce7e642192f95abac6

```
[100.00%] ··· bench_densearrsort.Unary.time_densesort                                                                                                              93.9±0.09μs
       before           after         ratio
     [eee12b6f]       [aaf8213d]
     <main>           <avx512-sort>
-     93.9±0.09μs      17.9±0.08μs     0.19  bench_densearrsort.Unary.time_densesort
```

r-devulap commented:
Looks like univeral intrinsic functions for reduce min/max (`npyv_reduce_min_u32`, etc) are defined locally in `loops_minmax.dispatch.c.src` which isn't ideal. Perhaps we should move those to  `numpy/core/src/common/simd/` ?

Commit:
MAINT: rename reduce_min/max functions to not clash with local def in loops_minmax.dispatch.c.src

goldsteinn commented:
> > Whats happening though is the same pivot is being returned each time when neither `smallest` nor `biggest` are updated. So its just a really expensive path to `heapsort` (and causes bad performance for dense ranges).
> 
> Do you have an example of an array where this worsens performance? At least for the example you provided in https://github.com/goldsteinn/possible-off-by-one/blob/master/test.c, this patch still speeds up quicksort by nearly 5x. See the benchmark: [r-devulap@d80c7b3](https://github.com/r-devulap/numpy/commit/d80c7b3a042c92f5773b366aa652a99f5a3a2094)
> 
> ```
> [100.00%] ··· bench_densearrsort.Unary.time_densesort                                                                                                              93.9±0.09μs
>        before           after         ratio
>      [eee12b6f]       [aaf8213d]
>      <main>           <avx512-sort>
> -     93.9±0.09μs      17.9±0.08μs     0.19  bench_densearrsort.Unary.time_densesort
> ```

Big fan of the patch. The implementation is awesome, was just comparing it to some other similar implementations and saw the performance issue.

I updated my repo with a benchmark. Added a small change to make pivot selection depend on `max_iters` so it wouldn't loop and get ~10-20% performance.

Either way nice patch :) 

https://github.com/goldsteinn/possible-off-by-one

```
$> ./build.sh && ./test
avx512_qsort : 34161.681
avx512_qsort2: 28359.471
```

r-devulap commented:
> Big fan of the patch. The implementation is awesome, was just comparing it to some other similar implementations and saw the performance issue.

Mind sharing what the other similar implementations are? Just curious to see. 

> I updated my repo with a benchmark. Added a small change to make pivot selection depend on `max_iters` so it wouldn't loop and get ~10-20% performance.

Oh that is nice. I like the idea of adapting your pivot selection, but have to be careful with: `get_pivot_epu32(arr, left + (max_iters % 8), right - (max_iters % 8));` The get_pivot function doesn't check if `left < right`. That can be easily accommodated though.

goldsteinn commented:
@r-devulap 

Something like
https://github.com/damageboy/VxSort/blob/master/README.md
or even just radix sort.

goldsteinn commented on numpy/core/src/npysort/qsort-32bit-avx512.h.src:
```diff
@@ -0,0 +1,518 @@
+#if @FIXNAN@
+        npy_intp nan_count = replace_nan_with_inf(arr, arrsize);
+#endif
+        qsort_@type@(arr, 0, arrsize-1, 2*log2(arrsize));
```
Maybe replace `log2` with `(63 - __builtin_clzll(arrsize))` (`lzcntq`). AFAIK all machines that support `avx512` have fast `__builtin_clzll` and the precision from `log2` isn't really needed.

mattip commented:
@r-devulap do you intend on updating this or should we put it in as-is and the pivot can be changed in a subsequent PR?

r-devulap commented:
@mattip We can put this as is. I will have to run some benchmarks and test out the suggestions to pivot, will add it as a separate PR.
